native/dsteqr.go

```diff
@@ -0,0 +1,386 @@
+// d contains the eigenvalues in ascending order. d must have length n and
+// Dsteqr will panic otherwise.
+//
+// e, on entry, contains the off-diagonal elements of the tridiagonal matrix on
+// entry, and is overwritten during the call to Dsteqr. e must have length n-1 and
+// Dsteqr will panic otherwise.
+//
+// z, on entry, contains the n×n orthogonal matrix used in the reduction to
+// tridiagonal form if compz == lapack.EigBoth. On exit, if
+// compz == lapack.EigBoth, z contains the orthonormal eigenvectors of the
+// original symmetric matrix, and if compz == lapack.EigDecomp, z contains the
+// orthonormal eigenvectors of the symmetric tridiagonal matrix. z is not used
+// if compz == lapack.EigValueOnly.
+//
+// work must have length at least max(1, 2*n-2) if the eigenvectors are computed,
+// and Dsteqr will panic otherwise.
+func (impl Implementation) Dsteqr(compz lapack.EigComp, n int, d, e, z []float64, ldz int, work []float64) (ok bool) {
+	if len(d) < n {
+		panic(badD)
+	}
+	if len(e) < n-1 {
+		panic(badE)
+	}
+	if compz != lapack.EigValueOnly && compz != lapack.EigBoth && compz != lapack.EigDecomp {
+		panic(badEigComp)
+	}
+	if compz != lapack.EigValueOnly {
+		if len(work) < max(1, 2*n-2) {
+			panic(badWork)
+		}
+		checkMatrix(n, n, z, ldz)
+	}
+
+	var icompz int
+	if compz == lapack.EigDecomp {
+		icompz = 1
+	} else if compz == lapack.EigBoth {
+		icompz = 2
+	}
+
+	if n == 0 {
+		return true
+	}
+	if n == 1 {
+		if icompz == 2 {
+			z[0] = 1
+		}
+		return true
+	}
+
+	bi := blas64.Implementation()
+
+	eps := dlamchE
+	eps2 := eps * eps
+	safmin := dlamchS
+	safmax := 1 / safmin
+	ssfmax := math.Sqrt(safmax) / 3
+	ssfmin := math.Sqrt(safmin) / eps2
+
+	// Compute the eigenvalues and eigenvectors of the tridiagonal matrix.
+	if icompz == 2 {
+		impl.Dlaset(blas.All, n, n, 0, 1, z, ldz)
+	}
+	const maxit = 30
+	nmaxit := n * maxit
+
+	jtot := 0
+
+	// Determine where the matrix splits and choose QL or QR iteration for each
+	// block, according to whether top or bottom diagonal element is smaller.
+	// TODO(btracey): Reformat the gotos to use structural flow techniques.
+	// TODO(btracey): Move these variables closer to actual usage when
+	// gotos are removed.
+	l1 := 0
+	nm1 := n - 1
+	var iscale, l, lend, lendsv, lm1, lsv, m, mm1 int
+	var anorm, c, g, p, r, rt1, rt2, s float64
+
+Ten:
+	if l1 > n-1 {
+		goto OneSixty
+	}
+	if l1 > 0 {
+		e[l1-1] = 0
+	}
+	if l1 <= nm1 {
+		for m = l1; m < nm1; m++ {
+			test := math.Abs(e[m])
+			if test == 0 {
+				goto Thirty
+			}
+			if test <= (math.Sqrt(math.Abs(d[m]))*math.Sqrt(math.Abs(d[m+1])))*eps {
+				e[m] = 0
+				goto Thirty
+			}
+		}
+	}
+	m = n - 1
+Thirty:
+	l = l1
+	lsv = l
+	lend = m
+	lendsv = lend
+	l1 = m + 1
+	if lend == 0 {
+		goto Ten
+	}
+
+	// Scale submatrix in rows and columns L to Lend
+	anorm = impl.Dlanst(lapack.MaxAbs, lend-l+1, d[l:], e[l:])
+	if anorm == 0 {
+		goto Ten
+	}
+	// TODO(btracey): Replace this with a constant
```

Comment:
Sorry, it was a note to myself to have the same up/down constants. I replaced this.